libraries/lib-time-and-pitch/FormantShifter.cpp

```diff
@@ -0,0 +1,183 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  FormantShifter.cpp
+
+  Matthieu Hodgkinson
+
+**********************************************************************/
+#include "FormantShifter.h"
+#include "FormantShifterLoggerInterface.h"
+#include <algorithm>
+#include <fstream>
+
+namespace
+{
+constexpr auto MapToPositiveHalfIndex(int index, int fullSize)
+{
+   const auto inputIsValid = fullSize > 0 && fullSize % 2 == 0;
+   if (!inputIsValid)
+      return 0;
+   if (index >= 0)
+      index = index % fullSize;
+   else
+      index = fullSize - (-index % fullSize);
+   if (index > fullSize / 2)
+      index = fullSize - index;
+   return index;
+}
+
+static_assert(MapToPositiveHalfIndex(-7, 4) == 1);
+static_assert(MapToPositiveHalfIndex(9, 4) == 1);
+
+constexpr float GetLog2(float x)
+{
+   // https://stackoverflow.com/a/28730362
+   // Precise up to two decimal places. If given unnormalized power, this
+   // shouldn't be a problem.
+   union
+   {
+      float val;
+      int32_t x;
+   } u = { x };
+   auto log_2 = (float)(((u.x >> 23) & 255) - 128);
+   u.x &= ~(255 << 23);
+   u.x += 127 << 23;
+   log_2 += ((-0.3358287811f) * u.val + 2.0f) * u.val - 0.65871759316667f;
+   return log_2;
+}
+
+// `x` has length `fftSize/2+1`.
+// Returns the last bin that wasn't zeroed.
+size_t ResampleFreqDomain(float* x, size_t fftSize, double factor)
+{
+   const auto size = fftSize / 2 + 1;
+   const auto end = std::min(size, size_t(size * factor));
+   std::vector<float> tmp(end);
+   for (size_t i = 0; i < end; ++i)
+   {
+      const int int_pos = i / factor;
+      const float frac_pos = 1.f * i / factor - int_pos;
+      const auto k = MapToPositiveHalfIndex(int_pos, fftSize);
+      const auto l = MapToPositiveHalfIndex(int_pos + 1, fftSize);
+      tmp[i] = (1 - frac_pos) * x[k] + frac_pos * x[l];
+   }
+   std::copy(tmp.begin(), tmp.begin() + end, x);
+   if (end < size)
+      std::fill(x + end, x + size, 0.f);
+   return end;
+}
+} // namespace
+
+FormantShifter::FormantShifter(
+   int sampleRate, double cutoffQuefrency,
+   FormantShifterLoggerInterface& logger)
+    : cutoffQuefrency { cutoffQuefrency }
+    , mSampleRate { sampleRate }
+    , mLogger { logger }
+{
+}
+
+void FormantShifter::Reset(size_t fftSize)
+{
+   mFft = std::make_unique<staffpad::audio::FourierTransform>(fftSize);
+   const auto numBins = fftSize / 2 + 1;
+   mEnvelope.setSize(1, numBins);
+   mCepstrum.setSize(1, fftSize);
+   mEnvelopeReal.resize(numBins);
+   mWeights.resize(numBins);
+}
+
+void FormantShifter::Reset()
+{
+   mFft.reset();
+}
+
+void FormantShifter::Process(
+   const float* powSpec, std::complex<float>* spec, double factor)
```

Comment:
As you use `factor` for calculating the bounds of your iterators, it's maybe a good idea to check `factor` is in expected range.